### PR TITLE
Consistency in decision forest + v1 namespace

### PR DESCRIPTION
**Pull request by rlnx**
- Fix order of template arguments in decision forest 
- Add v1 namespace for DF and some base functionality
- Move `empty_deleter` and other classes in `dal/memory.hpp` to `dal::detail`

Wanted to add `v1` to all algorithms at once, but too many changes in DF, decided to open separate PR

### CODE REVIEW

File: cpp/oneapi/dal/algo/decision_forest/backend/model_impl.hpp [hunk 1]
```diff
@@ -17,23 +17,36 @@
...
     }
+
+    std::int64_t tree_count = 0;
+    std::int64_t class_count = 0;
+
+private:
+    backend::model_interop* interop_ = nullptr;
 };
 
+using model_impl_cls = detail::model_impl<task::classification>;
```
Comment (michael-smirnov) [inline cpp/oneapi/dal/algo/decision_forest/backend/model_impl.hpp:49]: 1. detail::v1::model_impl?
2. why we need these aliases in public headers, public namespace of the library?

Comment (rlnx) [reply cpp/oneapi/dal/algo/decision_forest/backend/model_impl.hpp:49]: It's not public, moreover it's in backend.

**Review by michael-smirnov — APPROVED** [review]

**Review by KulikovNikita — APPROVED** [review]

File: cpp/oneapi/dal/algo/decision_forest/infer_types.hpp [hunk 2]
```diff
@@ -21,27 +21,35 @@
 namespace oneapi::dal::decision_forest {
 
 namespace detail {
-template <typename Task = task::by_default>
+namespace v1 {
+template <typename Task>
```
[the remark this answers is not in the document]
Comment (rlnx) [reply cpp/oneapi/dal/algo/decision_forest/infer_types.hpp:25]: It's forward declaration of impl class, I'd not use default template parameters here, need to apply to other impls

**Pull request merged**